tests/unit/test_openstack.py

```diff
@@ -180,7 +206,164 @@
         ],
         net_pools,
         no_usable_traceback_no_ips_left,
-        ValidationError("Error: no available networks for 3 hosts with IPv4"),
+        ValidationError("no available networks"),
+        "mrack.conf",
+        "allow",
+        75,
+    ),
+    (
+        "test3-hosts_force_spread_no_usable_traceback_no_ips_left",
```

Comment:
I'm not sure if I'm able to do the review of this. The way the input data is written is very difficult to understand what it is actually testing and if it behaves correctly. I have no idea what are the avail abilities in each of the test cases and it is difficult to calculate in head.

I'd expect a comment for each test saying what it is actually testing. The test name tries to encode it, but it its difficult to decode. An actual sentence would be helpful. 

You also don't need to add all of the cases into single test method. It's possible to create multiple tests which are using some common logic implemented in different helper functions and thus save the number of input args and make the test usable. 

E.g. when testing spreading and no spreading, the most important attributes are number of host and utilization of the network - these are the values to be displayed in the input params.

Generate the network availability structure with a helper method. The input can be list of dicts/tripples with `network_name`,  `total_ips`, `used_ips`.

Reply:
now the tests should look better

Reply:
thanks for the commit cherrypicked and now i know what you meant.